File: Backend/app/pw_library/python_methods/vandermonde.py
import pandas as pd

from Backend.app.pw_library.python_methods.gaussian_elimination import gaussian_elimination_verbose


def vandermonde_linear_regression(x, y, degree=1):
    """
    Performs polynomial regression using the Vandermonde matrix method.

    Parameters:
    x (list of float): Input values.
    y (list of float): Output values.
    degree (int): Degree of the polynomial (1 for linear regression).

    Returns:
    list of float: Coefficients of the polynomial in ascending order.
    """
    # Number of data points
    n = len(x)

    # Construct the Vandermonde matrix
    V = [[x[i] ** j for j in range(degree + 1)] for i in range(n)]

    # Compute (V^T * V) and (V^T * y)
    V_transpose = [[V[j][i] for j in range(n)] for i in range(degree + 1)]
    VTV = [[sum(V_transpose[i][k] * V[k][j] for k in range(n)) for j in range(degree + 1)] for i in range(degree + 1)]
    VTy = [sum(V_transpose[i][k] * y[k] for k in range(n)) for i in range(degree + 1)]

    # Display VTV as a DataFrame
    VTV_df = pd.DataFrame(VTV, columns=[f'a{j}' for j in range(degree + 1)], index=[f'a{i}' for i in range(degree + 1)])

    # Display VTy as part of the system
    VTy_df = pd.DataFrame(VTy, index=[f'a{i}' for i in range(degree + 1)], columns=['b'])

    # Solve the system VTV * a = VTy for the coefficients a
    a = gaussian_elimination_verbose(VTV, VTy)

    return a, VTV_df, VTy_df

File: Backend/app/pw_library/rust_methods/vandermonde.rs
use nalgebra::{DMatrix, DVector};

fn vandermonde_linear_regression(x: &[f64], y: &[f64], degree: usize) -> Result<(DVector<f64>, DMatrix<f64>, DVector<f64>), &'static str> {
    // Number of data points
    let n = x.len();

    // Construct the Vandermonde matrix
    let mut v = DMatrix::zeros(n, degree + 1);
    for i in 0..n {
        for j in 0..=degree {
            v[(i, j)] = x[i].powi(j as i32);
        }
    }

    // Compute V^T * V and V^T * y
    let v_transpose = v.transpose();
    let vtv = &v_transpose * &v;
    let vty = &v_transpose * DVector::from_column_slice(y);

    // Solve the system VTV * a = VTy for the coefficients a
    let a = vtv.lu().solve(&vty).ok_or("Gaussian elimination failed: Matrix may be singular")?;

    Ok((a, vtv, vty))
}
